dotCMS/src/main/java/com/dotcms/rendering/velocity/services/ContentletLoader.java

```diff
@@ -0,0 +1,969 @@
+     * otherwise you should use the other method {@link #invalidateLive(Contentlet) invalidateLive}
+     * where you only pass the contentlet
+     * </p>
+     *
+     * @param contentlet
+     * @param identifier of the contentlet, if its null then it will do a search using contentlet
+     *        parameter
+     * @throws DotDataException
+     * @throws DotSecurityException
+     */
+    public void invalidateLive(Contentlet contentlet, Identifier identifier) throws DotDataException, DotSecurityException {
+        Identifier localIdentifier = (identifier == null) ? APILocator.getIdentifierAPI()
+            .find(contentlet) : identifier;
+
+
+        invalidate(contentlet, localIdentifier, false);
+
+
+    }
+
+    /**
+     * Invalidate contentlet that is working
+     *
+     * @param contentlet
+     * @throws DotDataException
+     * @throws DotSecurityException
+     */
+    public void invalidateWorking(Contentlet contentlet) throws DotDataException, DotSecurityException {
+        invalidateWorking(contentlet, null);
+    }
+
+    /**
+     * Invalidate contentlet that is working.
+     * <p>
+     * NOTE: This method is helpful when you would like to save calls finding the identifier,
+     * otherwise you should use the other method {@link #invalidateWorking(Contentlet)
+     * invalidateWorking} where you only pass the contentlet
+     * </p>
+     *
+     * @param contentlet
+     * @param identifier of the contentlet, if its null then it will do a search using contentlet
+     *        parameter
+     * @throws DotDataException
+     * @throws DotSecurityException
+     */
+    public void invalidateWorking(Contentlet contentlet, Identifier identifier) throws DotDataException, DotSecurityException {
+        Identifier localIdentifier = (identifier == null) ? APILocator.getIdentifierAPI()
+            .find(contentlet) : identifier;
+
+
+        invalidate(contentlet, localIdentifier, true);
+
+    }
+
+    private void invalidate(final Contentlet content, final Identifier identifier, final boolean EDIT_MODE)
+            throws DotDataException, DotSecurityException {
+
+        removeContentletFile(content, identifier, EDIT_MODE);
+
+        if (content.getContentType()
+            .baseType() == BaseContentType.HTMLPAGE) {
+            new PageLoader().removePageFile(APILocator.getHTMLPageAssetAPI()
+                .fromContentlet(content), identifier, EDIT_MODE);
+        }
+        if (content != null && content.isVanityUrl()) {
+            // remove from cache
+            VanityUrlServices.getInstance()
+                .invalidateVanityUrl(content);
+        }
+        if (content != null && content.isKeyValue()) {
+            // remove from cache
+            CacheLocator.getKeyValueCache()
+                .remove(content);
+        }
+    }
+
+    public InputStream buildVelocity(Contentlet content, Identifier identifier, boolean EDIT_MODE, String filePath)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 140 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') This method has 556 lines, which is greater than the 75 lines authorized. Split it into smaller methods. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS138)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Rename this local variable to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00117)